### ISSUE TITLE
Document C API for sample traversal?

### ISSUE THREAD

**Issue by molpopgen** [tskit-dev/tskit#394]
The documentation in #392 looks great.  If folks are using that part of the API, then they probably want to know about accessing the number of samples below a node and the samples themselves.  The latter is noodly enough that it may warrant its own function instead of documenting the loop?

**Comment by jeromekelleher** [comment]
I'm not sure about documented the sample access code (i.e., the ``left_sample`` etc stuff). It's not clear there's any performance benefit over a straightforward traversal loop, unless you're revisiting the samples lots of times. IIRC, genotype decoding is a bit faster if you use a traversal (annoyingly!). So, I'm not sure there's much point in tying ourselves to this functionality in the long-term if it's not actually that useful.

I agree on sample counts all right though, that's a fundamental operation.

**Comment by molpopgen** [comment]
Sounds good.  

Regarding the genotype decoding:  does "a bit" refer to the case if iterating over trees *with* sample list updating or *without*?  If the former, then is it possible that "a  bit" could be "a lot" because the updating the sample lists itself from tree to tree is relatively costly?

**Comment by jeromekelleher** [comment]
> Regarding the genotype decoding: does "a bit" refer to the case if iterating over trees with sample list updating or without? If the former, then is it possible that "a bit" could be "a lot" because the updating the sample lists itself from tree to tree is relatively costly?

Yes, it's the total cost of decoding the genotypes, including tree generation step. It seems like the cost of maintaining the sample lists for *every* node outweighs gains that we make when actually iterating over the samples. But, hmm, I probably only checked this for a given rho/mu ratio. Presumably, when the mutation rate is high enough, the relative advantages might change.

Perhaps we should document the sample list stuff; it may be useful for some applications and it is more convenient than writing a traversal loop.

**Comment by jeromekelleher** [comment]
I'm not too excited about documenting this for 1.0 @molpopgen - I still feel like it could be done better and I don't want to tie us to the details of the current implementation for all time. Any objections to dropping it for 1.0?

**Comment by molpopgen** [comment]
No problem.
